### PR TITLE
Drill down scale down metrics

### PR DESCRIPTION
**Pull request by bskiba**
Split scale down duration into three parts:
1. Find nodes to remove
2. Node deletion
3. Misc operations

### CODE REVIEW

**Comment by bskiba** [comment]
I'm still testing how this is going to work, but would appreciate your comments on whether the metrics make sense.

File: cluster-autoscaler/core/scale_down.go [hunk 1]
```diff
@@ -266,6 +268,11 @@ func (sd *ScaleDown) TryToScaleDown(nodes []*apiv1.Node, pods []*apiv1.Pod, pdbs
 	// to recreate on other nodes.
 	emptyNodes := getEmptyNodes(candidates, pods, sd.context.MaxEmptyBulkDelete, sd.context.CloudProvider)
 	if len(emptyNodes) > 0 {
+		nodeDeletionStart := time.Now()
+		defer func(nodeDeletionStart time.Time, nodeDeletionDuration *time.Duration) {
+			stop := time.Now()
+			*nodeDeletionDuration = stop.Sub(nodeDeletionStart)
+		}(nodeDeletionStart, &nodeDeletionDuration)
```
Comment (MaciekPytel) [inline cluster-autoscaler/core/scale_down.go]: That part feels a bit hacky. There aren't actually any returns from this branch, until the deletion is already done. Why not just update nodeDeletionDuration in line 314 (before if finalError == nil)

File: cluster-autoscaler/core/static_autoscaler.go [hunk 2]
```diff
@@ -285,7 +285,7 @@ func (a *StaticAutoscaler) RunOnce(currentTime time.Time) errors.AutoscalerError
 			scaleDownStart := time.Now()
 			metrics.UpdateLastTime("scaleDown", scaleDownStart)
 			result, typedErr := scaleDown.TryToScaleDown(allNodes, allScheduled, pdbs)
-			metrics.UpdateDuration("scaleDown", scaleDownStart)
```
Comment (MaciekPytel) [inline cluster-autoscaler/core/static_autoscaler.go]: I would consider completely removing 'scaleDown' label. I don't think there is any case where we want to look at node deletion + calculations together.

File: cluster-autoscaler/metrics/metrics.go [hunk 3]
```diff
@@ -50,6 +50,12 @@ const (
 	LogLongDurationThreshold = 5 * time.Second
 )
 
+const (
```
Comment (MaciekPytel) [inline cluster-autoscaler/metrics/metrics.go]: Why another const block 2 lines after an existing one?

File: cluster-autoscaler/metrics/metrics.go [hunk 4]
```diff
@@ -50,6 +50,12 @@ const (
...
+	ScaleDownNodeDeletion      = "scaleDown:nodeDeletion"
+	ScaleDownFindNodesToRemove = "scaleDown:findNodesToRemove"
+	ScaleDownMiscOperations    = "scaleDown:miscOperations"
```
Comment (MaciekPytel) [inline cluster-autoscaler/metrics/metrics.go]: If you want to do it nicely maybe put all function names as consts here and give them a type (similar to NodeScaleDownReason above)? Though this is more of a drive-by cleanup and not really part of this PR, so don't feel obliged to do it.

**Comment by MaciekPytel** [comment]
Added some comments, but overall I think this is a very good way to get useful metrics about scale-down. :+1:

File: cluster-autoscaler/core/scale_down.go (hunk 1 above)
Comment (bskiba) [inline cluster-autoscaler/core/scale_down.go]: Only reason was it felt a little scary to spread the duration start and stop in the code so much as it will be very easy to put a little return between them and mess with the setup. But I guess hacking around this might not be the best way.

Comment (MaciekPytel) [inline cluster-autoscaler/core/scale_down.go]: Good point. I think the best solution here would be to move deleting empty nodes to a separate function (similar to how we have deleteNode function). Than we get to keep the start and stop close to each other without resorting to hacks.

As a side benefit that would slightly reduce the pain I feel every time I read our central functions, like RunOnce, TryToScaleDown or ScaleUp and once again see how bloated and hard to follow they are.

Comment (bskiba) [inline cluster-autoscaler/core/scale_down.go]: Factored out into two separate functions.

File: cluster-autoscaler/core/static_autoscaler.go (hunk 2 above)
Comment (bskiba) [inline cluster-autoscaler/core/static_autoscaler.go]: Maybe just leave it for sanity checking? But no strong feelings, I can remove it.

File: cluster-autoscaler/metrics/metrics.go (hunk 4 above)
Comment (bskiba) [inline cluster-autoscaler/metrics/metrics.go]: Extracted all the currently existing labels I think. Let me know if this makes sense.

File: cluster-autoscaler/metrics/metrics.go (hunk 3 above)
Comment (bskiba) [inline cluster-autoscaler/metrics/metrics.go]: Right now to be able to document the whole block of consts at once. (unless there's another way to do it)

File: cluster-autoscaler/core/scale_down.go [hunk 5]
```diff
@@ -266,55 +268,23 @@ func (sd *ScaleDown) TryToScaleDown(nodes []*apiv1.Node, pods []*apiv1.Pod, pdbs
...
-					glog.Errorf("Problem with empty node deletion: %v", err)
-					finalError = err
-				}
-			case <-time.After(timeLeft):
-				finalError = errors.NewAutoscalerError(errors.TransientError, "Failed to delete nodes in time")
-			}
-		}
-		if finalError == nil {
+		sd.scheduleDeleteEmptyNodes(emptyNodes, readinessMap, confirmation)
+		err := waitForEmptyNodesDeleted(emptyNodes, confirmation)
```
Comment (MaciekPytel) [inline cluster-autoscaler/core/scale_down.go]: I don't like the fact that scheduleDeleteEmptyNodes is a method and waitForEmptyNodesDeleted is a function. I understand it doesn't need to access any fields on sd, but still I would make it a method for consistency.

File: cluster-autoscaler/metrics/metrics.go [hunk 6]
```diff
@@ -50,10 +54,19 @@ const (
...
-	ScaleDownNodeDeletion      = "scaleDown:nodeDeletion"
-	ScaleDownFindNodesToRemove = "scaleDown:findNodesToRemove"
-	ScaleDownMiscOperations    = "scaleDown:miscOperations"
+	ScaleDown                  FunctionLabel = "scaleDown"
+	ScaleDownNodeDeletion      FunctionLabel = "scaleDown:nodeDeletion"
+	ScaleDownFindNodesToRemove FunctionLabel = "scaleDown:findNodesToRemove"
+	ScaleDownMiscOperations    FunctionLabel = "scaleDown:miscOperations"
+	ScaleUp                    FunctionLabel = "scaleUp"
+	FindUnneeded               FunctionLabel = "findUnneeded"
+	UpdateClusterStateLabel    FunctionLabel = "updateClusterState"
```
Comment (MaciekPytel) [inline cluster-autoscaler/metrics/metrics.go]: Why does this one has word label in name and others don't?

File: cluster-autoscaler/metrics/metrics.go (hunk 3 above)
Comment (MaciekPytel) [inline cluster-autoscaler/metrics/metrics.go]: Oh, and that actually passes the linter? Cool, I didn't know that, will start doing the same :)

File: cluster-autoscaler/metrics/metrics.go (hunk 6 above)
Comment (bskiba) [inline cluster-autoscaler/metrics/metrics.go]: It clashes with a function name in this file, but I don't want to put the Label suffix on all the names. Couldn't think of anything clever.

File: cluster-autoscaler/core/scale_down.go (hunk 5 above)
Comment (bskiba) [inline cluster-autoscaler/core/scale_down.go]: Done

File: cluster-autoscaler/metrics/metrics.go (hunk 6 above)
Comment (bskiba) [inline cluster-autoscaler/metrics/metrics.go]: Renamed to UpdateState
